Src/Plugins/Security/SEC101_011.SquareCredentialsValidator.cs

```diff
@@ -43,5 +60,74 @@ protected override IEnumerable<ValidationResult> IsValidStaticHelper(Dictionary<
 
             return new[] { validationResult };
         }
+
+        protected override ValidationState IsValidDynamicHelper(ref Fingerprint fingerprint,
+                                                                ref string message,
+                                                                Dictionary<string, string> options,
+                                                                ref ResultLevelKind resultLevelKind)
+        {
+            string id = fingerprint.Id;
+            string secret = fingerprint.Secret;
+
+            const string codeForRequest = "123";
+            const string uri = "https://connect.squareup.com/oauth2/token";
+            const string codeNotFoundMessage = "Authorization code not found";
+
+            try
+            {
+                HttpClient client = CreateHttpClient();
+
+                var dict = new Dictionary<string, string>()
+                {
+                    { "client_id", id },
+                    { "code", codeForRequest },
+                    { "client_secret", secret },
+                    { "grant_type", "authorization_code" },
+                };
+
+                using var request = new HttpRequestMessage(HttpMethod.Post, uri);
+                request.Content = new FormUrlEncodedContent(dict);
+
+                using HttpResponseMessage response = client
+                    .SendAsync(request, HttpCompletionOption.ResponseHeadersRead)
+                    .GetAwaiter()
+                    .GetResult();
+
+                switch (response.StatusCode)
+                {
+                    case HttpStatusCode.OK:
+                    {
+                        // This should theoretically never be hit, a "code" of 123 is much too short.
+                        // But if we do get a 200, it's a valid Credential.
+                        return ValidationState.Authorized;
+                    }
```

Comment:
You need to have the courage of your comment's convictions here. If this code should never be hit, you should be throwing an exception or throwing an error condition. 

Since I agree with you this code shouldn't be hit, it seems like any execution through this code path, in fact, would be a bug and not flag an actual find.

In general, it doesn't help us to code defensively. Instead, aggressively fail unexpected code paths and our test process (if we are testing properly) should flush those out.

Reply:
My specific suggestion is to use unexpected response code for this code path. Because that's what it is, right?

Reply:
Makes sense, just going to remove the block entirely and we can have it fall to 'Unexpected Response Code'